### PR TITLE
feat(microservices) add matchMqttPattern to server mqtt

### PR DESCRIPTION
**Pull request by hiepthai**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #2948

## What is the new behavior?
Add wildcard support for MQTT topics and return related handlers

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

File: packages/microservices/server/server-mqtt.ts [hunk 1]
```diff
@@ -66,6 +70,46 @@ export class ServerMqtt extends Server implements CustomTransportStrategy {
...
+        return false;
+    }
+
+    return patternLength === topicLength;
+  }
+
+  public getHandlerByPattern(pattern: string): MessageHandler | null {
+    const route = this.getRouteFromPattern(pattern);
+
+    for (const [key, value] of this.messageHandlers) {
```
Comment (kamilmysliwiec) [inline packages/microservices/server/server-mqtt.ts]: This change would have a huge impact on performance. On each message, we'd have to iterate through all existing handlers + call `matchMqttPattern` (multiple string-related operations)

Comment (hiepthai) [reply packages/microservices/server/server-mqtt.ts]: Hi, thank for your quick response. I will find a more effective way to match these pattern.

Comment (izio38) [reply packages/microservices/server/server-mqtt.ts]: That's how `mqtt-pattern` package [is doing it](https://github.com/RangerMauve/mqtt-pattern/blob/5f76bab764af69655ef1c54c1461f8aed921749a/index.js#L18).

Maybe we could check if the first part of the topic pattern is the same (before "argument" resolution)?

File: packages/microservices/server/server-mqtt.ts [hunk 2]
```diff
@@ -66,6 +70,46 @@ export class ServerMqtt extends Server implements CustomTransportStrategy {
...
+      if (!currentTopic && currentPattern !== MQTT_WILDCARD_ALL) return false;
+
+      if (patternChar === MQTT_WILDCARD_ALL) return i === lastIndex;
+
+      if (
+        patternChar !== MQTT_WILDCARD_SINGLE &&
+        currentPattern !== currentTopic
+      )
+        return false;
+    }
```
Comment (kamilmysliwiec) [inline packages/microservices/server/server-mqtt.ts]: Before you perform any pattern matching, you should simply test if patterns are equal. Then, check if `pattern` has either `MQTT_WILDCARD_SINGLE` or `MQTT_WILDCARD_ALL`

Comment (hiepthai) [reply packages/microservices/server/server-mqtt.ts]: Hi, I added some conditions before calling `matchMqttPattern` method.

**Comment by kamilmysliwiec** [comment]
I'm not entirely sure if that's a good idea to add custom features on top of the transport strategy. The mqtt library should support such a functionality out-of-the-box.

**Comment by frederikprijck** [comment]
@kamilmysliwiec I stumbled upon this PR because I'm also in need of a way to be able to do this.
Actualy, mqtt supports listening to all topics. But Nest doesn't support this yet.

Here's a quick walkthrough of my findings:

```
@EventPattern('#')
  async handlePresence2(data: Record<string, unknown>) {
    console.log(data);
  }
```

Using this succesfuly subscribs to '#' ( = all topics) here: https://github.com/nestjs/nest/blob/master/packages/microservices/server/server-mqtt.ts#L56

When using the above wildcard, handleMessage (https://github.com/nestjs/nest/blob/master/packages/microservices/server/server-mqtt.ts#L78) is correctly triggered for every topic.
However, I get the following error when I have an event coming in:

> [Server] There is no matching event handler defined in the remote service.

I think this is because of this line: https://github.com/nestjs/nest/blob/master/packages/microservices/server/server-mqtt.ts#L96

As you can see, it's looking for the handler based on the channel name: https://github.com/nestjs/nest/blob/master/packages/microservices/server/server-mqtt.ts#L96

However, there is no explicit handler for this topic, so that's why handler is undefined and will be resulting in the error.

A way to solve this could be to replace this line https://github.com/nestjs/nest/blob/master/packages/microservices/server/server-mqtt.ts#L96 with
`const handler = this.getHandlerByPattern(pattern) || this.getHandlerByPattern('#');`

Probably we want to have a more robust solution, maybe we want to call both handlers incase we have one that matches the pattern and one that uses the wildcard etc...

**Comment by frederikprijck** [comment]
@kamilmysliwiec Sorry to ping you again, any update on this?

**Comment by EliyahuFisherman** [comment]
@kamilmysliwiec please review this ASAP, I can help with this if necessary

**Comment by hiepthai** [comment]
Here a quick and easy solution for everybody at the moment, please check this https://github.com/nestjs/nest/issues/2948#issuecomment-569899580

**Comment by Nosfistis** [comment]
> A way to solve this could be to replace this line https://github.com/nestjs/nest/blob/master/packages/microservices/server/server-mqtt.ts#L96 with
> `const handler = this.getHandlerByPattern(pattern) || this.getHandlerByPattern('#');`

There is the more complex case of '+' wildcard to add there!

> Probably we want to have a more robust solution, maybe we want to call both handlers incase we have one that matches the pattern and one that uses the wildcard etc...

I think that both handlers should be called. Having a "more specific" behavior contradicts the usage of the mqtt.js library.

**Comment by kamilmysliwiec** [comment]
Let's track this as part of this PR! Thank you @hiepthai!

**Pull request merged**